riscv/processor.cc

```diff
@@ -397,6 +397,33 @@ void state_t::reset(processor_t* const proc, reg_t max_isa)
   const reg_t henvcfg_init = (proc->extension_enabled(EXT_SVPBMT) ? HENVCFG_PBMTE : 0);
   csrmap[CSR_HENVCFG] = henvcfg = std::make_shared<henvcfg_csr_t>(proc, CSR_HENVCFG, henvcfg_mask, henvcfg_init, menvcfg);
 
+  const reg_t mstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? MSTATEEN0_FCSR : 0) |
+                               MSTATEEN0_CS | MSTATEEN0_HENVCFG | MSTATEEN_HSTATEEN;
+  const reg_t mstateen0h_mask = MSTATEEN0H_HENVCFG | MSTATEENH_HSTATEEN;
+  const reg_t hstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? HSTATEEN0_FCSR : 0) |
+                               HSTATEEN0_CS | HSTATEEN0_SENVCFG | HSTATEEN_SSTATEEN;
+  const reg_t hstateen0h_mask = HSTATEEN0H_SENVCFG | HSTATEENH_SSTATEEN;
+  const reg_t sstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? SSTATEEN0_FCSR : 0) |
+                               SSTATEEN0_CS;
+  for (int i = 0; i < 4; i++) {
+    const reg_t mstateen_mask = i == 0 ? mstateen0_mask : MSTATEEN_HSTATEEN;
+    csrmap[CSR_MSTATEEN0 + i] = mstateen[i] = std::make_shared<stateen_csr_t>(proc, CSR_MSTATEEN0 + i, mstateen_mask, 0);
+    if (xlen == 32) {
+      const reg_t mstateenh_mask = i == 0 ? mstateen0h_mask : MSTATEENH_HSTATEEN;
+      csrmap[CSR_MSTATEEN0H + i] = std::make_shared<rv32_high_csr_t>(proc, CSR_MSTATEEN0H + i, mstateenh_mask, mstateen[i]);
+    }
+
+    const reg_t hstateen_mask = i == 0 ? hstateen0_mask : HSTATEEN_SSTATEEN;
+    csrmap[CSR_HSTATEEN0 + i] = hstateen[i] = std::make_shared<hstateen_csr_t>(proc, CSR_HSTATEEN0 + i, hstateen_mask, 0, i);
+    if (xlen == 32) {
+      const reg_t hstateenh_mask =  i == 0 ? hstateen0h_mask : HSTATEENH_SSTATEEN;
+      csrmap[CSR_HSTATEEN0H + i] = std::make_shared<rv32_high_csr_t>(proc, CSR_HSTATEEN0H + i, hstateenh_mask, hstateen[i]);
+    }
```

Comment:
If hypervisor extension is not implemented, these CSRs should not exist.

This also means that `hstateen_csr_t::verify_permissions()` needs to dynamically check for the H extension, since `misa.H` is writable.

Reply:
> This also means that `hstateen_csr_t::verify_permissions()` needs to dynamically check for the H extension, since `misa.H` is writable.

OK. I'll add this.



> Edit: it's sufficient to do this check only inside `hstateen_csr_t::verify_permissions()`; no need to skip their creation here.

It's OK. However, I found the check for xlen is added here for other similar csrs.

Reply:
> If hypervisor extension is not implemented, these CSRs should not exist.
> 
> This also means that `hstateen_csr_t::verify_permissions()` needs to dynamically check for the H extension, since `misa.H` is writable.

This have been checked in the base class csr_t, so I think it's unnecessary here

```
void csr_t::verify_permissions(insn_t insn, bool write) const {
  // Check permissions. Raise virtual-instruction exception if V=1,
  // privileges are insufficient, and the CSR belongs to supervisor or
  // hypervisor. Raise illegal-instruction exception otherwise.
  unsigned priv = state->prv == PRV_S && !state->v ? PRV_HS : state->prv;

  if ((csr_priv == PRV_S && !proc->extension_enabled('S')) ||
      (csr_priv == PRV_HS && !proc->extension_enabled('H')))
    throw trap_illegal_instruction(insn.bits());
```

Reply:
Yes, you're right, since the CSR's address indicates privilege, this is already handled.